SSA/Projects/Scf/Scf.lean

```diff
@@ -0,0 +1,175 @@
+            exact i
+          }
+        let to_iterate := to_loop_iterate (α := ⟦_t⟧) f'
+        let loop_fn := niter.iterate (op := to_iterate)
+        (loop_fn (0, vstart)).2
+
+    | .iterate k, (.cons (x : BitVec 32) .nil), (.cons (f : _ → BitVec 32) .nil) =>
+      let f' (v :  BitVec 32) : BitVec 32 := f  (Ctxt.Valuation.nil.snoc v)
+      k.iterate f' x
+      -- let f_k := Nat.iterate f' k
+      -- f_k x
+
+def cst {Γ : Ctxt _} (n : ℤ) : Expr Op Γ .int  :=
+  Expr.mk
+    (op := .const n)
+    (ty_eq := rfl)
+    (args := .nil)
+    (regArgs := .nil)
+
+def add {Γ : Ctxt _} (e₁ e₂ : Var Γ .int) : Expr Op Γ .int :=
+  Expr.mk
+    (op := .add)
+    (ty_eq := rfl)
+    (args := .cons e₁ <| .cons e₂ .nil)
+    (regArgs := .nil)
+
+def iterate {Γ : Ctxt _} (k : Nat) (input : Var Γ Ty.int) (body : Com Op [.int] .int) : Expr Op Γ .int :=
+  Expr.mk
+    (op := .iterate k)
+    (ty_eq := rfl)
+    (args := .cons input .nil)
+    (regArgs := HVector.cons body HVector.nil)
+
+def if_ {Γ : Ctxt _} {t : Ty} (cond : Var Γ .bool) (v : Var Γ t) (then_ else_ : Com Op [t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .if t)
+    (ty_eq := rfl)
+    (args := .cons cond <| .cons v .nil)
+    (regArgs := HVector.cons then_ <| HVector.cons else_ <| HVector.nil)
+
+def run {Γ : Ctxt _} {t : Ty} (v : Var Γ t) (body : Com Op [t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .run t)
+    (ty_eq := rfl)
+    (args := .cons v .nil)
+    (regArgs := HVector.cons body <| HVector.nil)
+
+def for_ {Γ : Ctxt _} {t : Ty} (niter : Var Γ .nat) (v : Var Γ t) (body : Com Op [.nat, t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .for t)
+    (ty_eq := rfl)
+    (args := .cons niter <| .cons v .nil)
+    (regArgs := HVector.cons body <| HVector.nil)
+
+/-- 'if' condition of a true variable evaluates to the true region body. -/
+theorem if_true {t : Ty} (cond : Var Γ .bool) (hcond : Γv cond = true) (v : Var Γ t) (then_ else_ : Com Op [t] t) :
+  Expr.denote (ScfRegion.if_ (t := t) cond v then_ else_) Γv =
+  Expr.denote (ScfRegion.run (t := t) v then_) Γv := by
+    simp[Expr.denote, if_, run]
+    simp_peephole[hcond] at Γv
+    simp[ite_true]
+
+/-- 'if' condition of a true variable evaluates to the true region body. -/
+theorem if_false {t : Ty} (cond : Var Γ .bool) (hcond : Γv cond = false) (v : Var Γ t) (then_ else_ : Com Op [t] t) :
+  Expr.denote (ScfRegion.if_ (t := t) cond v then_ else_) Γv =
+  Expr.denote (ScfRegion.run (t := t) v else_) Γv := by
+    simp[Expr.denote, if_, run]
+    simp_peephole[hcond] at Γv
+    simp[ite_true]
+
+attribute [local simp] Ctxt.snoc
+
+/-- running `f(x) = x + x` 0 times is the identity. -/
+def lhs : Com Op [.int] .int :=
+  Com.lete (iterate (k := 0) ⟨0, by simp[Ctxt.snoc]⟩ (
+      Com.lete (add ⟨0, by simp[Ctxt.snoc]⟩ ⟨0, by simp[Ctxt.snoc]⟩) -- fun x => (x + x)
+      <| Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+  )) <|
+  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+
+def rhs : Com Op [.int] .int :=
+  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
```

Comment:
Is there a reason this is not defined in MLIR syntax?

Reply:
#128 is green. Is sth missing?

Reply:
What about this comment. At the very least, it would be nice to have an end-to-end rewrite that shows the MLIR syntax if the lhs/rhs things dont work.